src/control/server/storage/config.go

```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
 	return nil
 }
 
+type stringSet map[string]struct{}
+
+func (ss stringSet) Add(s string) error {
+	if _, exists := ss[s]; exists {
+		return errors.Errorf("duplicate device %q", s)
+	}
+	ss[s] = struct{}{}
+	return nil
+}
+
+func (ss stringSet) Strings() []string {
+	var out []string
+	for s := range ss {
+		out = append(out, s)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// BdevDeviceList represents a set of block device addresses.
+type BdevDeviceList struct {
+	// As this is the most common use case, we'll make the embedded type's methods
+	// available directly on the type.
+	hardware.PCIAddressSet
+
+	// As a fallback for non-PCI bdevs, maintain a map of strings.
+	stringBdevSet stringSet
+}
+
+// maybePCI does a quick check to see if a string could possibly be a PCI address.
+func maybePCI(addr string) bool {
+	comps := strings.Split(addr, ":")
+	if len(comps) != 3 {
+		return false
+	}
+	return (len(comps[0]) == 6 || len(comps[0]) == 4) && len(comps[1]) == 2 && len(comps[2]) >= 2
+}
+
+// fromStrings creates a BdevDeviceList from a list of strings.
+func (bdl *BdevDeviceList) fromStrings(addrs []string) error {
+	if bdl == nil {
+		return errors.New("nil BdevDeviceList")
+	}
+
+	if bdl.stringBdevSet == nil {
+		bdl.stringBdevSet = stringSet{}
+	}
+
+	for _, strAddr := range addrs {
+		if !maybePCI(strAddr) {
+			if err := bdl.stringBdevSet.Add(strAddr); err != nil {
+				return err
+			}
+			continue
+		}
+
+		addr, err := hardware.NewPCIAddress(strAddr)
+		if err != nil {
+			return err
+		}
+
+		if bdl.Contains(addr) {
+			return errors.Errorf("duplicate PCI address %s", addr)
+		}
+
+		if err := bdl.Add(addr); err != nil {
+			return err
+		}
+	}
+
+	if len(bdl.stringBdevSet) > 0 && bdl.PCIAddressSet.Len() > 0 {
+		return errors.New("cannot mix PCI and non-PCI block device addresses")
+	}
+
+	return nil
+}
+
+func (bdl *BdevDeviceList) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var tmp []string
+	if err := unmarshal(&tmp); err != nil {
```

Comment:
Probably a newbie golang question, but do not see how it could work without giving at least one more input parameter as for the json Unmarshall function.